packages/framework/ai-collab/src/explicit-strategy/agentEditReducer.ts

```diff
@@ -0,0 +1,499 @@
+): void {
+	if (typeof json === "object") {
+		if (json === null) {
+			return;
+		}
+		if (Array.isArray(json)) {
+			for (const element of json) {
+				populateDefaults(element, definitionMap);
+			}
+		} else {
+			assert(
+				typeof json[typeField] === "string",
+				"The typeField must be present in new JSON content",
+			);
+			const nodeSchema = definitionMap.get(json[typeField]);
+			assert(nodeSchema?.kind === NodeKind.Object, "Expected object schema");
+		}
+	}
+}
+
+function getSchemaIdentifier(content: TreeEditValue): string | undefined {
+	switch (typeof content) {
+		case "boolean": {
+			return booleanSchema.identifier;
+		}
+		case "number": {
+			return numberSchema.identifier;
+		}
+		case "string": {
+			return stringSchema.identifier;
+		}
+		case "object": {
+			if (content === null) {
+				return nullSchema.identifier;
+			}
+			if (Array.isArray(content)) {
+				throw new UsageError("Arrays are not currently supported in this context");
+			}
+			if (isFluidHandle(content)) {
+				return handleSchema.identifier;
+			}
+			return content[typeField];
+		}
+		default: {
+			throw new UsageError("Unsupported content type");
+		}
+	}
+}
+
+function contentWithIds(content: TreeNode, idGenerator: IdGenerator): TreeEditObject {
+	return JSON.parse(toDecoratedJson(idGenerator, content)) as TreeEditObject;
+}
+
+/**
+ * Manages applying the various types of {@link TreeEdit}'s to a a given {@link TreeNode}.
+ */
+export function applyAgentEdit(
+	treeNode: TreeNode,
+	treeEdit: TreeEdit,
+	idGenerator: IdGenerator,
+	definitionMap: ReadonlyMap<string, SimpleNodeSchema>,
+	validator?: (edit: TreeNode) => void,
+): TreeEdit {
+	objectIdsExist(treeEdit, idGenerator);
+	switch (treeEdit.type) {
+		case "insert": {
+			const { array, index } = getPlaceInfo(treeEdit.destination, idGenerator);
+
+			const parentNodeSchema = Tree.schema(array);
+			populateDefaults(treeEdit.content, definitionMap);
+
+			const schemaIdentifier = getSchemaIdentifier(treeEdit.content);
+
+			// We assume that the parentNode for inserts edits are guaranteed to be an arrayNode.
+			const allowedTypes = [
+				...normalizeAllowedTypes(parentNodeSchema.info as ImplicitAllowedTypes),
+			];
+
+			for (const allowedType of allowedTypes.values()) {
+				if (allowedType.identifier === schemaIdentifier && typeof allowedType === "function") {
+					const simpleNodeSchema = allowedType as unknown as new (dummy: unknown) => TreeNode;
+					const insertNode = new simpleNodeSchema(treeEdit.content);
+					validator?.(insertNode);
+					array.insertAt(index, insertNode);
+					return {
+						...treeEdit,
+						content: contentWithIds(insertNode, idGenerator),
+					};
+				}
+			}
+			fail("inserted node must be of an allowed type");
+		}
+		case "remove": {
+			const source = treeEdit.source;
+			if (isObjectTarget(source)) {
+				const node = getNodeFromTarget(source, idGenerator);
+				const parentNode = Tree.parent(node);
+				// Case for deleting rootNode
```

Comment:
If we don't have the `setRoot` operation, I don't believe there is any way to do this. I think we can just assert that the parent is defined.

Reply:
Agreed, removed the ability to set the root to undefined via a remove operation in commit bcf7e43a9a34491c64543194d98321fb9b54ef8d